### DISCUSSION TITLE
How to add coastlines to a UXARRAY plot

### DISCUSSION THREAD

**Discussion by bstephens82** [UXARRAY/uxarray#861] [Q&A] [answered]
Apologies if this is duplicating another discussion (I looked but didn't see one quite like what I was looking for), but I am just wondering how to include coastlines in uxarray plots.  In some maps I've seen it's pretty obvious where the continents are simply from looking at the contours, but I have some noisier data that's unfortunately not so clear.  I can plot a map, but without the coastlines it's hard to interpret.  Is there a simple fix for this?  Currently I'm using the command:
~~~
uxds['ug_mean_lev'][0,:].plot(title="ug_mean_lev", cmap='cubehelix_r', projection=ccrs.PlateCarree())
~~~

**Comment by philipc2** [comment, marked as answer]
Hi @bstephens82 

Since our plotting API is written using [GeoViews](https://geoviews.org/user_guide/Geometries.html), you can add any feature you'd like by using the following syntax

```Python
import geoviews.feature as gf
projection = projection=ccrs.PlateCarree()

uxds['ug_mean_lev'][0,:].plot(title="ug_mean_lev", cmap='cubehelix_r', projection=ccrs.PlateCarree()) * gf.coastline(projection=projection)
```

You can see an example of this in our [MPAS Visualization Example](https://uxarray.readthedocs.io/en/latest/examples/visualization/3_75km_mpas.html). I would note that only our Point plotting currently supports other projections as of right now using our direct API. We are working on getting this in soon.

If you are interested in a more classic Matplotlib implementation using Cartopy, we have recently added a new [User Guide](https://uxarray.readthedocs.io/en/latest/user-guide/mpl.html) section describing how to achieve this. This implementation does support any projection (even for Polygon plotting). 

Using Matplotlib, it would look something like:

```Python
import matplotlib.pyplot as plt
import cartopy.crs as ccrs
import cartopy.feature as cfeature

projection=ccrs.PlateCarree()

# create a Matplotlib PolyCollection
pc = uxds['ug_mean_lev'][0,:].to_polycollection(projection=projection)

pc.set_antialiased(False)
pc.set_cmap("cubehelix_r")

fig, ax = plt.subplots(
    1,
    1,
    figsize=(10, 5),
    facecolor="w",
    constrained_layout=True,
    subplot_kw=dict(projection=projection),
)

# add geographic features
ax.add_feature(cfeature.COASTLINE)
ax.add_feature(cfeature.BORDERS)

ax.add_collection(pc)
ax.set_global()
plt.title("ug_mean_lev")

**Comment by bstephens82** [comment]
Thanks very much @philipc2!  I got your first suggestion working.  Regarding your second suggestion, does that require a recent update to uxarray or something?  I get error messages when I try to run it, like it complains that "projection" is not a recognized argument for `to_polycollection`. Thanks again.

**Reply by philipc2** [reply]
It was added recently in our [v2024.07.0](https://github.com/UXARRAY/uxarray/releases/tag/v2024.07.0) release. 

It's available both on [PyPi ](https://pypi.org/project/uxarray/)and [Conda Forge](https://anaconda.org/conda-forge/uxarray)

**Comment by bstephens82** [comment]
Hi @philipc2 thanks again for your help so far, I've figured out those issues I think.  I had one other question...what if I have a variable on an unstructured grid that's not in the `uxds` dataset defined above, but calculated from a variable on an unstructured grid, and I want to plot that?  For example let's say I take a variable directly from the unstructured grid, but average it vertically so it's only 2D.  Now I can't use the uxds['variable-name'].plot() syntax...is there another way to do that using uxarray?  Thanks!

**Reply by philipc2** [reply]
> what if I have a variable on an unstructured grid that's not in the uxds dataset defined above, but calculated from a variable on an unstructured grid

If your calculation has so far been completely contained within the UXarray ecosystem, you should still easily be able to plot it as long as it still resides on an unstructured grid element.

For example: 
```Python
# 2m temp in kelvin
t2m = uxds['t2m']

# apply some operation to t2m (returns a UxDataArray
t2m_c = t2m - 273.15

# plot with the same syntax
t2m_c.plot()
```

You can verify that you're still working with UXarray by checking if your output has a `.uxgrid` attached to it.

I'd like to mention that the `uxds['variable-name'].plot()` syntax is only needed if you are working with a UxDataset or multi-dimensional DataArray.

Our plotting routines operate on 1-D slices of data mapped to unstructured grid elements.

Feel free to provide a minimal example if you'd like me to try it out!

**Reply by bstephens82** [reply]
Okay thanks I will try that and get back in touch if I can't get it working.  Also, is there a way to extract a variable using uxarray, but extract it onto a lat/lon grid?  For example we might have `uxds['U']` with dimensions (time, level, column)...is there a way to extract that variable into a numpy array with dimensions (time, level, lat, lon)?

**Reply by philipc2** [reply]
This is something we'd like to support (#726) 

I'll look into putting together an initial unstructured to structured method that uses a nearest neighbor approach, as that is something we already have the infrastructure for.

**Reply by bstephens82** [reply]
> > what if I have a variable on an unstructured grid that's not in the uxds dataset defined above, but calculated from a variable on an unstructured grid
> 
> If your calculation has so far been completely contained within the UXarray ecosystem, you should still easily be able to plot it as long as it still resides on an unstructured grid element.
> 
> For example:
> 
> ```python
> # 2m temp in kelvin
> t2m = uxds['t2m']
> 
> # apply some operation to t2m (returns a UxDataArray
> t2m_c = t2m - 273.15
> 
> # plot with the same syntax
> t2m_c.plot()
> ```
> 
> You can verify that you're still working with UXarray by checking if your output has a `.uxgrid` attached to it.
> 
> I'd like to mention that the `uxds['variable-name'].plot()` syntax is only needed if you are working with a UxDataset or multi-dimensional DataArray.
> 
> Our plotting routines operate on 1-D slices of data mapped to unstructured grid elements.
> 
> Feel free to provide a minimal example if you'd like me to try it out!

Hi @philipc2, I have a question related to this since I'm running into a surprising problem.  I'm trying to simply subtract two uxarray variables and the plot the result, and it's not working.  My code looks like this:
~~~
diff = uxds2['UPWP_UG_MIN_LEV'] - uxds2['UPWP_UG_MAX_LEV']
diff[0,:].plot(title="ug_mean_lev", projection=ccrs.PlateCarree()) * gf.coastline(projection=projection)
~~~
And the error message I get is this:
ValueError: NoneType type could not be interpreted as Dimension.  Dimensions must be declared as a string, tuple, dictionary or Dimension type.

The two arrays I'm subtracting have the same shape (30, 48600), as does the `diff` result.  The same plotting syntax works if I just try to plot one or the other of the original two variables alone, but for some reason it's not working on the result.  Any ideas?  Thanks!

**Reply by bstephens82** [reply]
@philipc2 Just checking back to see if there was an answer for this.  Also, and apologies for the noob question, but is there an easy way to plot multiple panels using uxarray?  Like plotting two or more maps on a single figure?  Thanks again.

**Reply by philipc2** [reply]
Hi @bstephens82 

I must have missed this on my GitHub notifications! Thanks for the extra ping.

Can you try doing `diff[0].plot(title="ug_mean_lev", projection=ccrs.PlateCarree()) * gf.coastline(projection=projection)`

IF you can show a screenshot of what the print out for `diff` looks like as well that would be helpfull.


> but is there an easy way to plot multiple panels using uxarray?

If you are using our direct plotting API (i.e. `.plot()`) , you can use the `+` operator between two plots to add them together. Since our plotting API is written around HoloViews, we use the same syntax that is shown [here](https://holoviews.org/user_guide/Composing_Elements.html). 


[Here is an example of two plots stacked on top of each other. ](https://uxarray.readthedocs.io/en/latest/user-guide/plotting.html#working-with-periodic-data)

I'd like to share our [Visualization Cookbook ](https://projectpythia.org/unstructured-grid-viz-cookbook/README.html)if you haven't seen it yet. We need to put a few updates in, but it should still be a good resource.

**Reply by bstephens82** [reply]
Thanks!  Using a "+" worked to get multiple plots, so that's great.  Regarding taking the difference, it looks like you're saying to try removing the ",:" after the "0", and I tried that but it didn't work.  The error message is the same.  `print(diff)` gives

<xarray.UxDataArray (time: 30, n_face: 48600)>
array([[  0.,   0.,   0., ...,   0., -11.,  -1.],
       [  0.,   0.,   0., ...,   0.,   0.,  -4.],
       [  0.,   0.,   0., ...,   0.,   0.,   0.],
       ...,
       [  0.,   0.,   0., ...,  -1.,  -3.,   0.],
       [  0.,   0.,   0., ...,  -1.,  -3.,   0.],
       [  0.,   0.,   0., ..., -47.,  -3.,   0.]], dtype=float32)
Coordinates:
  * time     (time) object 1979-01-01 00:30:00 ... 1979-01-01 15:00:00
Dimensions without coordinates: n_face

**Reply by philipc2** [reply]
@bstephens82 

There is a bug in our plotting API that requires each data array to have a name. I will work on fixing this. For the meantime, you can assign a name to `diff` using `diff.name='some name'`

**Reply by bstephens82** [reply]
I see, great, it's working now.  I will check out Holoviews, it seems like maybe that would help a lot because I've struggled with the syntax a lot for making plots this way.  I know matplotlib is another option, I will also be looking at that.  One last hopefully simple question if you don't mind--if I have a 2D field and want to plot it on a map, is there an easy way to limit the coutour limits?  For example the data in the plot might go from 0 to 100, is there an option to limit the contours from 25 to 50?  Thanks a lot again for all your help.

**Comment by bstephens82** [comment]
Hi @philipc2, I have another small question about something that doesn't seem to be working.  I would like to use the `.where()` functionality  with a uxds array, something like
~~~
variable = variable.where(variable<0,0)
~~~
and this doesn't give any complaints by itself, but when I try to plot the result of something like this, it gives a weird complaint about not recognizing the "title" I've put into the `.plot()` command.  Here is the code I'm trying to use:
~~~
upwp_ug = uxds['UPWP_UG_CLUBB']
upwp_ug = upwp_ug.where(upwp_ug<0,0)
#upwp_ug = abs(upwp_ug)
upwp_ug_sum=upwp_ug.sum(dim='ilev')
print(upwp_ug_sum.shape)
upwp_ug_sum.mean(dim='time').plot(title="3a1 minus dflt, sum(abs(UPWP_UG_CLUBB)) (m2/s2)", projection=ccrs.PlateCarree()) * gf.coastline(projection=projection)
~~~
If I comment out the line with `.where()` and uncomment the `abs()` value line, it plots fine.  But as is, it complains about the `title`:
~~~
AttributeError: Line2D.set() got an unexpected keyword argument 'title'
~~~
Any ideas?  Does `.where()` not work yet with UXARRAY?  Thanks!

**Reply by philipc2** [reply]
Looks like the return type is `xarray.DataArray`, which is not what we'd like. I'm creating an issue for this and will look into it.

**Comment by bstephens82** [comment]
Hi @philipc2, is there an easy way to get my uxarray plots on a logarithmic color scale?  I was able to figure out how to set limits to the color scale but not how to get it on a log scale.  Thanks!  Here's a piece of code as an example:
~~~
upwp_ug_sum.mean(dim='time').plot(title=f"sum", projection=ccrs.PlateCarree()) * gf.coastline(projection=projection)
~~~

**Reply by philipc2** [reply]
You'd like to set `logz=True`

<img width="886" alt="image" src="https://github.com/user-attachments/assets/fa99411e-306f-42a0-b40a-162f30a73472">

Here's a good reference guide for most supported parameters! 

https://hvplot.holoviz.org/user_guide/Customization.html
